packages/tailwindcss/src/compat/screens-config.ts

```diff
@@ -0,0 +1,94 @@
+import { rule } from '../ast'
+import type { DesignSystem } from '../design-system'
+import type { ResolvedConfig } from './config/types'
+
+export function registerScreensConfig(userConfig: ResolvedConfig, designSystem: DesignSystem) {
+  let screens = userConfig.theme.screens || {}
+
+  // We want to insert the breakpoints in the right order as best we can. In the
+  // core utility, all static breakpoint variants and the `min-*` functional
+  // variant are registered inside a group. Since all the variants within a
+  // group share the same order, we can use the always-defined `min-*` variant
+  // as the order.
+  let coreOrder = designSystem.variants.get('min')?.order ?? undefined
+
+  // Register static breakpoint variants for everything that comes from the user
+  // theme config.
+  for (let [name, value] of Object.entries(screens)) {
+    let coreVariant = designSystem.variants.get(name)
+
+    // Ignore it if there's a CSS value that takes precedence over the JS config
+    // and the static utilities are already registered.
+    //
+    // This happens when a `@theme { }` block is used that overwrites all JS
+    // config options. We rely on the resolution order of the Theme for
+    // resolving this. If Theme has a different value, we know that this is not
+    // coming from the JS plugin and thus we don't need to handle it explicitly.
+    let cssValue = designSystem.theme.resolveValue(name, ['--breakpoint'])
+    if (coreVariant && cssValue && !designSystem.theme.hasDefault(`--breakpoint-${name}`)) {
+      continue
+    }
+
+    let query: string | undefined
+    let insertOrder: number | undefined
+    if (typeof value === 'string') {
+      query = `(width >= ${value})`
+      insertOrder = coreOrder
+    } else if (typeof value === 'object' && value !== null) {
+      if (Array.isArray(value)) {
+        query = value.map(ruleForComplexScreenValue).join(', ')
+      } else {
+        query = ruleForComplexScreenValue(value) ?? ''
+        if ('min' in value && !('max' in value)) {
+          insertOrder = coreOrder
+        }
+      }
+    } else {
+      continue
+    }
+
+    if (coreOrder && insertOrder === undefined) {
+      insertOrder = allocateOrderAfter(designSystem, coreOrder)
```

Comment:
Is it possible for us to do this all at one time at the end rather than  after every inserted variant?

We're definitely doing more traversals and object replacements than we need to but if it proves to be too complex then this is fine for now.

Reply:
We have to do it once for every utility that is inserted this way. Each one will handed their own order. I do need a test for this since I think it might now be adding it in inverse insertion order.

Reply:
Yeah it would make sense that this is the reason some things are going being inserted in reverse order. I was thinking of something like this:

- Get a list of all the variants _after_ `min` (and their compare fns)
- Insert our appropriate screens into the `min` group
- Insert other screens after the `min` group
- Move the previously recorded variants past the last variant we just inserted (*only* if there is one)

I think this would work but I'm not sure if the separate compareFn map complicates that.

Reply:
Hm I’m not sure I follow, I think we can keep doing what we do right now but only loop over the variants in reverse order, this way we:

- Still insert appropriate screens into the `min` group. For that, the order does not matter since it'll use the compare function
- Insert complex variants _after_ the `min` group. But since we loop in reverse order, we will will always move the previously inserted complex variant down. 

I pushed a commit that changes this and it fixes the order.